src/reactor.rs

```diff
@@ -0,0 +1,741 @@
+                if let Some(step) = proposed_step.increment() {
+                    self.upgrade_step(keychain, moderator, height, round, *block_id, step)
+                        .await?;
+                }
+            }
+
+            Consensus::Reject => {
+                #[cfg(feature = "trace")]
+                tracing::trace!(
+                    "vote rejected, not enough validators for the round: height {}, round {}, author {:08x}, state: {:?}, validators: {}",
+                    height,
+                    round,
+                    validator,
+                    proposed_step,
+                    validators
+                );
+            }
+        }
+
+        #[cfg(feature = "trace")]
+        tracing::trace!(
+            "vote processed: height {}, round {}, author {:08x}, state: {:?}",
+            height,
+            round,
+            validator,
+            proposed_step
+        );
+
+        Ok(())
+    }
+
+    pub(crate) async fn receive_event(&mut self, _event: Event) {
+        #[cfg(feature = "trace")]
+        tracing::warn!("inbound events are not expected; ignored {:?}", _event);
+    }
+
+    pub(crate) async fn receive_notification<K, M>(
+        &mut self,
+        keychain: &K,
+        moderator: &mut M,
+        notification: Notification,
+    ) where
+        K: Keychain,
+        M: Moderator,
+    {
+        match notification {
+            Notification::Kill => self.should_quit = true,
+
+            Notification::NewValidator {
+                height,
+                validity,
+                validator,
+            } => self.add_validator(validator, height, validity),
+
+            Notification::Vote { vote } => {
+                if let Err(_e) = self.receive_vote(keychain, moderator, vote).await {
+                    #[cfg(feature = "trace")]
+                    tracing::error!("error receiving vote: {}", _e);
+                }
+            }
+
+            Notification::BlockAuthorized { height, block_id } => {
+                #[cfg(feature = "trace")]
+                tracing::debug!("block authorized for height {}", height);
+
+                self.metadata.authorize_block(block_id, height)
+            }
+
+            Notification::BlockProposeAuthorized { height, block_id } => {
+                #[cfg(feature = "trace")]
+                tracing::debug!("block propose authorized for height {}", height);
+
+                self.metadata.authorize_block_propose(height, block_id)
+            }
+        }
+    }
+
+    pub(crate) async fn receive_request<K, M>(
+        &mut self,
+        keychain: &K,
+        moderator: &mut M,
+        request: Request,
+    ) where
+        K: Keychain,
+        M: Moderator,
+    {
+        let response = match request {
+            Request::Commit { id, height, round } => Response::Commit {
+                id,
+                committed: self.commit(moderator, height, round).await,
+            },
```

Comment:
Hm, maybe I am missing something, but aren't `Request` send outside of bft? If they are, this `self.commit` does not do any checks for consensus but just jumps to next height, not sure if this is something we want.

Reply:
Yes, we accept this event to commit to a given height without checking the votes. Its useful to setup environments and/or re-sync a peer